filecoin-proofs/src/api/post_adapter.rs

```diff
@@ -0,0 +1,626 @@
+
+    let remainder = chunks.remainder();
+
+    for chunk in chunks {
+        fixed.push(GeneratePoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            challenge_seed: dynamic.challenge_seed,
+            input_parts: [chunk[0].clone(), chunk[1].clone()],
+        });
+    }
+
+    if !remainder.is_empty() {
+        let mut input_parts: [(Option<String>, Commitment); POST_SECTORS_COUNT] =
+            Default::default();
+
+        let vs = repeat_with(|| remainder[remainder.len() - 1].clone());
+        let rs = remainder.iter().cloned();
+        let it = rs.chain(vs).take(POST_SECTORS_COUNT).enumerate();
+
+        for (i, input_part) in it {
+            input_parts[i] = input_part;
+        }
+
+        fixed.push(GeneratePoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            challenge_seed: dynamic.challenge_seed,
+            input_parts,
+        });
+    }
+
+    fixed
+}
+
+/// Collapses return values from multiple fixed sectors-count PoSt generation
+/// operations into a single return value for dynamic sector count. This
+/// function discards any faults reported for duplicated CommRs.
+///
+pub fn generate_post_collect_output(
+    orig_comm_rs_len: usize,
+    xs: Vec<error::Result<GeneratePoStFixedSectorsCountOutput>>,
+) -> error::Result<GeneratePoStDynamicSectorsCountOutput> {
+    if xs.is_empty() {
+        return Err(format_err!("input vector must not be empty"));
+    }
+
+    let mut dynamic = GeneratePoStDynamicSectorsCountOutput {
+        proofs: Vec::new(),
+        faults: Vec::new(),
+    };
+
+    for (i, x) in xs.into_iter().enumerate() {
+        match x {
+            Err(e) => {
+                return Err(e);
+            }
+            Ok(fixed) => {
+                dynamic.proofs.push(fixed.proof);
+                for fault in fixed.faults.iter() {
+                    dynamic.faults.push(((i as u64) * 2) + fault)
+                }
+            }
+        }
+    }
+
+    dynamic.faults = dynamic.faults.into_iter().take(orig_comm_rs_len).collect();
+
+    Ok(dynamic)
+}
+
+/// Maps inputs for a single dynamic sectors-count PoSt verify operation to a
+/// vector of PoSt verify inputs for fixed sectors-count.
+///
+pub fn verify_post_spread_input(
+    dynamic: VerifyPoStDynamicSectorsCountInput,
+) -> error::Result<Vec<VerifyPoStFixedSectorsCountInput>> {
+    let faults_len = dynamic.faults.len();
+    let commrs_len = dynamic.comm_rs.len();
+    let proofs_len = dynamic.proofs.len();
+    let replicas_c = (commrs_len as f64 / POST_SECTORS_COUNT as f64)
+        .ceil()
+        .round() as usize;
+    let faults_max = dynamic.faults.iter().max();
+
+    if faults_len > commrs_len {
+        return Err(format_err!(
+            "LEN(faults) must <= LEN(comm_rs): {:?}, {:?}",
+            faults_len,
+            commrs_len
+        ));
+    }
+
+    if proofs_len != replicas_c {
```

Comment:
These types of checks (that don't trust the caller) might also be useful in the `generate_` counterparts of these functions.

Reply:
> These types of checks (that don't trust the caller) might also be useful in the generate_ counterparts of these functions.

I struggled to dream up invariants for the `generate_` calls that weren't already A) enforced by the type system or B) enforced at runtime by the Rust code I've already written. Can you think of any?